### PR TITLE
[PLGN-243] SentinelOne - Added new action: Quarantine Multiple

### PR DESCRIPTION
**Pull request by igorski-r7**
### Description

Describe the proposed changes:

  - Added new action: Quarantine Multiple

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [X] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [X] Screenshot of job output with the plugin changes
- [X] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [X] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [X] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [X] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [X] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [X] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [X] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [X] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [X] Work fully completed
- [X] Functional
  - [X] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [X] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [X] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [X] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [X] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `make validate` and make sure everything passes
2. Run the assessment tool: `icon-plugin run -A -R all -T all`. For single action validation: `icon-plugin run -A -R tests/my_action.json -T tests/my_action.json`
3. Copy (`icon-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

**Review by cmcnally-r7 — APPROVED** [review]

File: plugins/sentinelone/plugin.spec.yaml [hunk 1]
```diff
@@ -2088,40 +2103,39 @@ actions:
...
     output:
-      response:
-        title: Response
-        description: SentinelOne API call response data
-        type: quarantine_response
-        required: false
-        example: '{ "response": { "data": [ { "agentDomain": "WORKGROUP", "agentGroupId": "123123456712356789", "agentId": "123123456712356789" } } }'
+      successful:
+        title: Successful
+        description: List containing hosts with successful operation
```
Comment (dlaverty-r7) [inline plugins/sentinelone/plugin.spec.yaml:2129]: Suggested reword: 
List of hosts on which the operation was successfully attempted. (Note this does not guarantee that the operation itself was successful, only that the API call was able to be successfully performed)

File: plugins/sentinelone/plugin.spec.yaml [hunk 2]
```diff
@@ -2088,40 +2103,39 @@ actions:
...
+        type: "[]string"
+        required: true
+        example: '["hostname123", "hostname456"]'
+      failures:
+        title: Failure
+        description: List containing hosts with failure operation
```
Comment (dlaverty-r7) [inline plugins/sentinelone/plugin.spec.yaml:2135]: Suggested reword:
List of hosts on which the operation attempt was unsuccessful.

File: plugins/sentinelone/komand_sentinelone/actions/quarantine/action.py [hunk 3]
```diff
@@ -17,30 +19,40 @@ def __init__(self):
...
-
-        if self.__check_agents_found(agents):
-            self.logger.info(f"No agents found using the host information: {agent}.")
-            return {Output.RESPONSE: not_affected}
-
-        agent_obj = agents[0]
+        successful = []
+        failures = []
+        for agent in agents:
+            agents = self.connection.client.search_agents(agent, results_length=2)
```
Comment (dlaverty-r7) [inline plugins/sentinelone/komand_sentinelone/actions/quarantine/action.py:29]: Should this be reusing agents?

Comment (igorski-r7) [reply plugins/sentinelone/komand_sentinelone/actions/quarantine/action.py:29]: Great catch! It shouldn't 😄

**Review by llaszuk-r7 — APPROVED** [review]

File: plugins/sentinelone/komand_sentinelone/actions/quarantine_multiple/action.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                action_type = "connect"
+                if quarantine_state:
+                    if Helper.check_disconnected(agent_obj):
+                        self.logger.info(f"Agent: {agent} is already quarantined")
+                    if whitelist:
+                        Helper.find_in_whitelist(agent_obj, whitelist)
+                    action_type = "disconnect"
+                self.connection.agents_action(action_type, payload)
+                successful.append(agent)
+            except (PluginException, HTTPError) as error:
```
Comment (ablakley-r7) [inline plugins/sentinelone/komand_sentinelone/actions/quarantine_multiple/action.py:50]: I think this is a good way to handle agents who are already found quarantined. May wish to check with Connisoars on whether this is expected response however. As in, should they be added to success also?

File: plugins/sentinelone/komand_sentinelone/actions/quarantine_multiple/action.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+                continue
+
+            agent_obj = agents_founds[0]
+            payload = {"ids": [agent_obj["id"]]}
+
+            try:
+                action_type = "connect"
+                if quarantine_state:
+                    if Helper.check_disconnected(agent_obj):
+                        self.logger.info(f"Agent: {agent} is already quarantined")
```
Comment (ablakley-r7) [inline plugins/sentinelone/komand_sentinelone/actions/quarantine_multiple/action.py:44]: Can we reword this to include quarantine/unquarantine.

File: plugins/sentinelone/komand_sentinelone/actions/quarantine_multiple/action.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+                    if whitelist:
+                        Helper.find_in_whitelist(agent_obj, whitelist)
+                    action_type = "disconnect"
+                self.connection.agents_action(action_type, payload)
+                successful.append(agent)
+            except (PluginException, HTTPError) as error:
+                failures.append(Helper.return_failure_details(agent, str(error)))
+        return {
+            Output.SUCCESSFUL: successful,
```
Comment (ablakley-r7) [inline plugins/sentinelone/komand_sentinelone/actions/quarantine_multiple/action.py:53]: Can we rename 'successful' to 'completed'?

File: plugins/sentinelone/komand_sentinelone/actions/quarantine_multiple/action.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        # START INPUT BINDING - DO NOT REMOVE - ANY INPUTS BELOW WILL UPDATE WITH YOUR PLUGIN SPEC AFTER REGENERATION
+        agents = list(set(params.get(Input.AGENTS, [])))
+        whitelist = params.get(Input.WHITELIST, None)
+        quarantine_state = params.get(Input.QUARANTINE_STATE)
+        # END INPUT BINDING - DO NOT REMOVE
+
+        successful = []
+        failures = []
+        for agent in agents:
+            agents_founds = self.connection.client.search_agents(agent, results_length=2)
```
Comment (dlaverty-r7) [inline plugins/sentinelone/komand_sentinelone/actions/quarantine_multiple/action.py:29]: Reword: agents_founds -> agents_found

File: plugins/sentinelone/komand_sentinelone/actions/quarantine_multiple/action.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+                failures.append(Helper.return_failure_details(agent, error))
+                continue
+
+            agent_obj = agents_founds[0]
+            payload = {"ids": [agent_obj["id"]]}
+
+            try:
+                action_type = "connect"
+                if quarantine_state:
+                    if Helper.check_disconnected(agent_obj):
```
Comment (ablakley-r7) [inline plugins/sentinelone/komand_sentinelone/actions/quarantine_multiple/action.py:43]: Do we do any similar check if the quarantine_state is the other way around?

File: plugins/sentinelone/unit_test/test_quarantine_multiple.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+class TestQuarantineMultiple(TestCase):
+    @classmethod
+    @patch("requests.post", side_effect=Util.mocked_requests_get)
+    def setUpClass(cls, mock_request: Mock) -> None:
+        cls.action = Util.default_connector(QuarantineMultiple())
+
+    @patch("requests.request", side_effect=Util.mocked_requests_get)
+    @patch("requests.get", side_effect=Util.mocked_requests_get)
+    def test_should_success(self, mock_request: Mock, mock_get: Mock) -> None:
```
Comment (dlaverty-r7) [inline plugins/sentinelone/unit_test/test_quarantine_multiple.py:23]: Reword-> 'test_should_succeed'

**Review by ablakley-r7 — CHANGES_REQUESTED** [review]
Looks great! Have some comments on output naming and handling whitelist and success array.

File: plugins/sentinelone/unit_test/test_quarantine_multiple.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        expected = {
+            Output.SUCCESSFUL: ["hostname123"],
```
Comment (dlaverty-r7) [inline plugins/sentinelone/unit_test/test_quarantine_multiple.py:25]: As this is to check a multiple quarantine action, can you add a test to check for multiple as well as one hostname?

**Review by dlaverty-r7 — APPROVED** [review]
Generally lgtm, just a couple of small suggestions.

File: plugins/sentinelone/komand_sentinelone/actions/quarantine_multiple/action.py (hunk 4 above)
Comment (igorski-r7) [reply plugins/sentinelone/komand_sentinelone/actions/quarantine_multiple/action.py:53]: Hi 😄 Talked with them and thats fine to append `completed` array with agents who are already found quarantined 😄

**Review by ablakley-r7 — APPROVED** [review]

**Review by dlaverty-r7 — APPROVED** [review]
